### ISSUE TITLE
[MNT] Adress pandas deprecation of `M` frequency in datetimeindexes

### ISSUE THREAD

**Issue by benHeid** [sktime/sktime#6245]
During testing, I have seen a lot of warnings about the deprecation of the monthly `M` frequency. It should be replaced by `ME`.

<img width="968" alt="image" src="https://github.com/sktime/sktime/assets/9725296/2916d22e-458f-4f46-be59-aa2e5f120919">

**Comment by benHeid** [comment]
@fkiraly @yarnabrina I think this is strongly related to how our forecasting horizon is implemented. Since this warning is often raised in _fh.py line 415 and _fh.py line 172. A test that triggers this warning is: `test_hierarchical_with_exogeneous`


I think we have several options here:

* Remove the frequency from our forecasting horizon
* Support only PeriodAliases -> to_absolute would return PeriodIndex like
* Support only Frequencies -> to_absolute would return  DatetimeIndex like
* Try to support both and handle it flexible 
* Try to just fix it for now

I am currently in favor for a more general solution than just a fix. Since I suppose that a simple fix would just increase the code complexity and we had this discussion about reworking the ForecastnigHorizon several times.

**Comment by yarnabrina** [comment]
> I am currently in favor for a more general solution than just a fix.

Strongly agreed.

> Remove the frequency from our forecasting horizon

Assuming this means using range index for everything and not use the dates in the dataset, I'll be in favour of that. I am practically using the same myself in office work and managing dates to indices conversion myself.

(It will cause problem in generating dates for future though, so that will be the tricky part to reduce breaking changes I think.)

**Comment by fkiraly** [comment]
> Assuming this means using range index for everything and not use the dates in the dataset, I'll be in favour of that.

I'd be against that, handling irregularly spaced time series is an important feature for some of our users.

I'd be in favour of decoupling the `ForecastingHorizon` as much as possible from `pandas` frequency handling, e.g., use range or float index internally and do conversions in input and output points.

Possibly, internally everything we currently support is convertible to int:

* periods by mapping to "period since" convention
* time stamps by mapping to the internal integer that it uses anyway

**Comment by yarnabrina** [comment]
> use range or float index internally and do conversions in input and output points

May be I am missing something, but in my brain this is exactly same as what I assumed above.

**Comment by fkiraly** [comment]
> May be I am missing something, but in my brain this is exactly same as what I assumed above.

Perhaps - in your mental model, in the end design, is `ForecastingHorizon` able to produce datetime and period indices with `freq`, e.g., in `to_absolute`?

**Comment by benHeid** [comment]
@fkiraly you are referring to a IntIndex not RangeIndex internally or?

**Comment by fkiraly** [comment]
yes, because `RangeIndex` is always contiguous. Imo one needs to be able to represent sth like 0, 1, 2, 3, 5, etc as well, for irregularly spaced series or horizons.

**Comment by fkiraly** [comment]
@benHeid, @yarnabrina, do we have a plan on how to address this? Any suggestions?

**Comment by fkiraly** [comment]
Posting some related issues and a bug report in `pandas`:
https://github.com/sktime/sktime/pull/6574
https://github.com/sktime/sktime/issues/6572

`pandas` bug report: https://github.com/pandas-dev/pandas/issues/58974

**Comment by benHeid** [comment]
I think we should try to come back to this issue again... 

Should we try to convert it always to an IntIndex internally and save everything so that we can get the originally passed information?

**Comment by fkiraly** [comment]
yes, I would say so.

* periodic -> integer, with some convention on how to map onto date periodicities (e.g., complete periods since start of 1970; or offset to 1970 start plus incremental offsets)
* non-periodic -> float or integer, again with a convention on how to map. I suspect most cases will actually be convertible to integer (incl datetime), so I'm not sure if we need float at all.

**Comment by benHeid** [comment]
> * periodic -> integer, with some convention on how to map onto date periodicities (e.g., complete periods since start of 1970; or offset to 1970 start plus incremental offsets)
> * non-periodic -> float or integer, again with a convention on how to map. I suspect most cases will actually be convertible to integer (incl datetime), so I'm not sure if we need float at all.

Make sense

> I would even say, inside the `ForecastingHorizon` there should be no `pandas` at all in the end state, this state of affairs has just caused tears in the past. We should adopt a basic performant array structure, perhaps a simple 1D `numpy` array.

Removing pandas completely could be hard, due to the delegation or? 

Thus, would it be better to everything stepwise? 
1. Convert everything to `IntIndex` or 1D `numpy`array
2. Get rid off the remaining pandas functionality.

**Comment by fkiraly** [comment]
> Removing pandas completely could be hard, due to the delegation or?

Yes, but we can build it back step-by-step - going to integer index is perhaps a good intermediate state.

**Comment by benHeid** [comment]
I am creating a list for type and how to convert, please mention if I forgot an fh type that is supported:

* `pd.RangeIndex`
   * Conversion using `astype(int)` 
* `pd.DatetimeIndex`
   * Conversion using `astype(int)` 
* `pd.PeriodIndex`
   * Conversion using `astype("datetime64[ns]").astype(int)` 
   * I am unsure if this working for all period indexes... 
* `pd.TimedeltaIndex`
   *  **Problem** Conversion using `astype(int)` would lead to nanoseconds since cutoff date. However, we are not able to know the cutoff date to get the values since 1970. Thus, the meaning of the internal used integer would now differ.
   * Required passing of a cutoff date would probably be a breaking change.
* `pd.Index`
  * Conversion using `astype(int)` 
* `int`
  * `[int]` 
* `range`
   * `list(range)` 
* `timedelta`
   * **Problem** Same as for the `pd.TimedeltaIndex`
* `integer array`
  * Nothing todo. 
* `timedelta array
  * **Problem** Sam as for the `pd.TimedeltaIndex`

**Comment by fkiraly** [comment]
Looks complete!

I also wanted to add that in the `TrendForecaster` there already is a partial such coercion, here it is:

```python
def _get_X_numpy_int_from_pandas(x):
    """Convert pandas index to an sklearn compatible X, 2D np.ndarray, int type."""
    if isinstance(x, (pd.DatetimeIndex)):
        x = x.astype("int64") / 864e11
    else:
        x = x.astype("int64")
    return x.to_numpy().reshape(-1, 1)
```

Observations:

* `astype("int64")`, not `astype(int)`
* for `datetime` we divide through 864e11, which changes the type to `float` and the unit from nanosecond to day since start of 1970
* notably, for `PeriodIndex`, the unit remains number of periods

**Comment by fkiraly** [comment]
Regarding the `timedelta` problem, we could ensure that the encoding is consistent with `datetime`, and it results in a `is_relative=True` horizon.

**Comment by benHeid** [comment]
Also important for the reimplementation is, how should we test? Can we rely on that if the tests are green that the new forecasting horizon is working? 
Or should we add some extra regression tests just for this reimplementation, even if I am not sure on how to realize such regression tests...

**Comment by fkiraly** [comment]
> Also important for the reimplementation is, how should we test? Can we rely on that if the tests are green that the new forecasting horizon is working?

I think there are sufficient tests for external usage, but we will need to test internal logic carefully. I am also not sure how well all the public methods are currently covered, for the different cases - for instance, we discovered a while ago that the "exotic" frequencies like ME or 2H were not tested at the same level of coverage as the more common ones.

I would also think very carefully about the tests we need to ensure good coverage.

Finally, the `ForecastingHorizon` semi-inherits from `Index`, which may cause problems if we do not think about how to handle that - though in the first step (move to integer index) it should not pose problems yet.

**Comment by olofahlen** [comment]
Hi, I just started using sktime today and I am very pumped to dive into this project.

I'm not sure where best to report this, but I might have found a bug related to the frequency strings. Starting with this working example from the quick start guide:

```python
from sktime.datasets import load_airline
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.theta import ThetaForecaster
from sktime.performance_metrics.forecasting import mean_absolute_percentage_error
from sktime.split import temporal_train_test_split

y = load_airline()
y_train, y_test = temporal_train_test_split(y)

# Uncommenting the below will make it fail
# y = y.set_axis(y.index.to_timestamp(how="end").normalize())  # Converts periods like 1949-01 to timestamps like 1949-01-31

fh = ForecastingHorizon(y_test.index, is_relative=False)
forecaster = ThetaForecaster(sp=12)  # monthly seasonal periodicity
forecaster.fit(y_train)
y_pred = forecaster.predict(fh)
mean_absolute_percentage_error(y_test, y_pred)
```

If I uncomment the line above which converts to `pd.DatetimeIndex` then fitting fails with
```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File pandas/_libs/tslibs/timedeltas.pyx:732, in pandas._libs.tslibs.timedeltas.parse_timedelta_unit()

KeyError: 'me'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
File c:\Users\OlofAhlen\Projects\strategy-framework\.venv\Lib\site-packages\sktime\utils\datetime.py:52, in _coerce_duration_to_int(duration, freq)
     51 if isinstance(duration, pd.Timedelta):
---> 52     return int(duration / pd.Timedelta(count, unit))
     53 if isinstance(duration, pd.TimedeltaIndex):

File pandas/_libs/tslibs/timedeltas.pyx:1895, in pandas._libs.tslibs.timedeltas.Timedelta.__new__()

File pandas/_libs/tslibs/timedeltas.pyx:734, in pandas._libs.tslibs.timedeltas.parse_timedelta_unit()

ValueError: invalid unit abbreviation: ME

During handling of the above exception, another exception occurred:
```
and
```python
File c:\Users\OlofAhlen\Projects\strategy-framework\.venv\Lib\site-packages\sktime\utils\datetime.py:56, in _coerce_duration_to_int(duration, freq)
     54             return (duration / pd.Timedelta(count, unit)).astype(int)
     55     except ValueError:
---> 56         raise ValueError(
     57             "Index type not supported. Please consider using pd.PeriodIndex."
     58         )
     59 else:
     60     raise TypeError("`duration` type not understood.")

ValueError: Index type not supported. Please consider using pd.PeriodIndex.
```

Run the example above for full stack trace.

However, it works if I instead use

```python
forecaster = NaiveForecaster()
```

So it seems related to `ThetaForecaster` (and possibly some others).

I will be using sktime in a large time series framework I have developed myself that revolves around `pd.DatetimeIndex` (not `pd.PeriodIndex`), so my hope was that `pd.DatetimeIndex` would be fully supported. Also, as @fkiraly pointed out above, irregular frequencies (like `CustomBusinessDay`) play a very important role in my research.
